src/pybind/mgr/dashboard_v2/models/nodb.py

```diff
@@ -0,0 +1,892 @@
+
+    def contribute_to_class(self, cls, name):
+        clone = copy.deepcopy(self)
+        clone.model = cls
+        setattr(cls, name, clone)
+
+    def copy_manager_to(self, new_class):
+        new_class.add_to_class('objects', self)
+
+    def get_queryset(self):
+        return NodbQuerySet(self.model, context=NodbManager.nodb_context)
+
+    def all(self):
+        return self.get_queryset().all()
+
+    def get(self, **kwargs):
+        return self.get_queryset().get(**kwargs)
+
+    def filter(self, *args, **kwargs):
+        return self.get_queryset().filter(*args, **kwargs)
+
+
+class LazyProperty(object):
+    """
+    See also: django.db.models.query_utils.DeferredAttribute
+
+    Internally used by @bulk_attribute_setter().
+    """
+    class QuerySetPointer(object):
+        def __init__(self, target):
+            self.target = target
+
+    def __init__(self, field_name, eval_func, catch_exceptions, field_names):
+        self.field_name = field_name
+        self.eval_func = eval_func
+        self.catch_exceptions = catch_exceptions
+        self.field_names = field_names
+
+    def __get__(self, instance, owner=None):
+        """
+        runs eval_func which fills some lazy properties.
+        """
+        if hasattr(instance, '_query_set_pointer'):
+            query_set = instance._query_set_pointer.target
+        else:
+            # Fallback. Needed for objects without a queryset.
+            query_set = [instance]
+        if self.field_name in instance.__dict__:
+            return instance.__dict__[self.field_name]
+
+        if self.catch_exceptions is None:
+            self.eval_func(instance, query_set, self.field_names)
+        else:
+            try:
+                self.eval_func(instance, query_set, self.field_names)
+            except self.catch_exceptions as e:
+                logger.exception('failed to populate Field "{}" of {} ({})'
+                                 .format(self.field_name, str(instance), instance.__class__))
+                fields = instance.__class__.make_model_args({}, fields_force_none=self.field_names)
+                for field_name, value in list(fields.items()):
+                    setattr(instance, field_name, value)
+
+        if self.field_name not in instance.__dict__:
+            raise KeyError(
+                'LazyProperty: {} did not set {} of {}'.format(self.eval_func, self.field_name,
+                                                               instance))
+        return instance.__dict__[self.field_name]
+
+    def __set__(self, instance, value):
+        """
+        Deferred loading attributes can be set normally (which means there will
+        never be a database lookup involved.
+        """
+        instance.__dict__[self.field_name] = value
+
+
+def bulk_attribute_setter(field_names, catch_exceptions=None):
+    """
+    The idea @behind bulk_attribute_setter is to delay expensive calls to librados, until someone
```

Comment:
Do you really want the '@' to be in front of 'behind'?

Reply:
fixed